xray/utils/analyzermanager.go

```diff
@@ -203,27 +215,111 @@ func RemoveDuplicateValues(stringSlice []string) []string {
+	file := ExtractRelativePath(GetResultFileName(result), workingDir)
+	lineCol := GetResultLocationInFile(result)
+	text := *result.Message.Text
+	// Already exists
+	if index := GetResultIfExists(file, lineCol, text, *results); index >= 0 {
+		return index
+	}
+	// New result
+	newResult := SourceCodeScanResult{
+		Severity: GetResultSeverity(result),
+		SourceCodeLocation: SourceCodeLocation{
+			File:       file,
+			LineColumn: lineCol,
+			Text:       text,
+		},
+		Type: *result.RuleID,
+	}
+	index := len(*results)
+	*results = append(*results, newResult)
+
+	return index
+}
+
 func GetResultFileName(result *sarif.Result) string {
 	if len(result.Locations) > 0 {
-		filePath := result.Locations[0].PhysicalLocation.ArtifactLocation.URI
-		if filePath != nil {
-			return *filePath
-		}
+		return getResultFileName(result.Locations[0])
+	}
+	return ""
+}
+
+func getResultFileName(location *sarif.Location) string {
+	filePath := location.PhysicalLocation.ArtifactLocation.URI
+	if filePath != nil {
+		return *filePath
 	}
 	return ""
 }
 
 func GetResultLocationInFile(result *sarif.Result) string {
 	if len(result.Locations) > 0 {
-		startLine := result.Locations[0].PhysicalLocation.Region.StartLine
-		startColumn := result.Locations[0].PhysicalLocation.Region.StartColumn
-		if startLine != nil && startColumn != nil {
-			return strconv.Itoa(*startLine) + ":" + strconv.Itoa(*startColumn)
-		}
+		return getResultLocationInFile(result.Locations[0])
 	}
 	return ""
 }
 
+func getResultLocationInFile(location *sarif.Location) string {
+	startLine := location.PhysicalLocation.Region.StartLine
+	startColumn := location.PhysicalLocation.Region.StartColumn
+	if startLine != nil && startColumn != nil {
+		return strconv.Itoa(*startLine) + ":" + strconv.Itoa(*startColumn)
+	}
+	return ""
+}
+
+func GetResultLocationSnippet(location *sarif.Location) string {
+	if location != nil && location.PhysicalLocation != nil && location.PhysicalLocation.Region != nil && location.PhysicalLocation.Region.Snippet != nil {
+		return *location.PhysicalLocation.Region.Snippet.Text
+	}
+	return ""
+}
+
+func GetResultCodeFlows(result *sarif.Result, workingDir string) (flows [][]SourceCodeLocation) {
+	if len(result.CodeFlows) == 0 {
+		return
+	}
+	for _, codeFlow := range result.CodeFlows {
+		if codeFlow == nil || len(codeFlow.ThreadFlows) == 0 {
+			continue
+		}
+		for _, threadFlow := range codeFlow.ThreadFlows {
+			if threadFlow == nil || len(threadFlow.Locations) == 0 {
+				continue
+			}
+			flow := []SourceCodeLocation{}
+			for _, location := range threadFlow.Locations {
+				if location == nil {
+					continue
+				}
+				flow = append(flow, SourceCodeLocation{
+					File:       ExtractRelativePath(getResultFileName(location.Location), workingDir),
+					LineColumn: getResultLocationInFile(location.Location),
+					Text:       GetResultLocationSnippet(location.Location),
+				})
+			}
+			if len(flow) == 0 {
+				continue
+			}
+			flows = append(flows, flow)
```

Comment:
Let's extract this logic into a function for readability.
```suggestion
			flow := []SourceCodeLocation{}
			for _, location := range threadFlow.Locations {
				if location == nil {
					continue
				}
				flow = append(flow, SourceCodeLocation{
					File:       ExtractRelativePath(getResultFileName(location.Location), workingDir),
					LineColumn: getResultLocationInFile(location.Location),
					Text:       GetResultLocationSnippet(location.Location),
				})
			}
			if len(flow) > 0 {
				flows = append(flows, flow)
			}
			
```